`test_get_blocked_guids` intermittent test failure

In Travis-CI:

```
src/olympia/blocklist/tests/test_mlbf.py:66: in test_get_blocked_guids

    assert len(blocked_guids) == 5

E   AssertionError: assert 6 == 5
E     -6
E     +5
```

--- a/src/olympia/blocklist/cron.py
+++ b/src/olympia/blocklist/cron.py
@@ -7,7 +7,8 @@ from olympia.zadmin.models import get_config
 
 from .mlbf import MLBF
 from .models import Block
-from .tasks import MLBF_TIME_CONFIG_KEY, upload_filter_to_kinto
+from .tasks import (
+    MLBF_BASE_ID_CONFIG_KEY, MLBF_TIME_CONFIG_KEY, upload_filter_to_kinto)
 
 log = olympia.core.logger.getLogger('z.cron')
 
@@ -39,10 +40,26 @@ def upload_mlbf_to_kinto():
     generation_time = int(time.time() * 1000)
     mlbf = MLBF(generation_time)
     mlbf.generate_and_write_mlbf()
-    if last_generation_time:
+
+    base_filter_id = get_config(MLBF_BASE_ID_CONFIG_KEY, 0, json_value=True)
+    base_filter = MLBF(base_filter_id) if base_filter_id else None
+    make_base_filter = (
+        not base_filter or mlbf.should_reset_base_filter(base_filter))
+    if last_generation_time and not make_base_filter:
         try:
-            mlbf.write_stash(last_generation_time)
+            # optimize for when the base_filter was the previous generation so
+            # we don't have to load the blocked JSON file twice.
+            previous_mlbf = (
+                MLBF(last_generation_time)
+                if last_generation_time != base_filter_id else
+                base_filter)
+            mlbf.write_stash(previous_mlbf)
         except FileNotFoundError:
             log.info('No previous blocked.json so we can\'t create a stash.')
+            # fallback to creating a new base if stash fails
+            make_base_filter = True
 
-    upload_filter_to_kinto.delay(generation_time)
+    upload_filter_to_kinto.delay(
+        generation_time,
+        is_base=make_base_filter,
+        upload_stash=not make_base_filter)

--- a/src/olympia/blocklist/mlbf.py
+++ b/src/olympia/blocklist/mlbf.py
@@ -137,6 +137,11 @@ class MLBF():
         return os.path.join(
             settings.MLBF_STORAGE_PATH, self.id, 'stash.json')
 
+    @cached_property
+    def stash_json(self):
+        with storage.open(self.stash_path, 'r') as json_file:
+            return json.load(json_file)
+
     def generate_and_write_mlbf(self, *, blocked=None, not_blocked=None):
         stats = {}
 
@@ -186,7 +191,7 @@ class MLBF():
         # compare previous with current blocks
         extras, deletes = self.generate_diffs(
             previous_mlbf.blocked_json, self.blocked_json)
-        stash = {
+        self.stash_json = {
             'blocked': list(extras),
             'unblocked': list(deletes),
         }
@@ -194,7 +199,7 @@ class MLBF():
         stash_path = self.stash_path
         with storage.open(stash_path, 'w') as json_file:
             log.info("Writing to file {}".format(stash_path))
-            json.dump(stash, json_file)
+            json.dump(self.stash_json, json_file)
 
     def should_reset_base_filter(self, previous_base_mlbf):
         try:

--- a/src/olympia/blocklist/mlbf.py
+++ b/src/olympia/blocklist/mlbf.py
@@ -6,6 +6,7 @@ from collections import defaultdict
 
 from django.conf import settings
 from django.core.files.storage import default_storage as storage
+from django.utils.functional import cached_property
 
 from filtercascade import FilterCascade
 from filtercascade.fileformats import HashAlgorithm
@@ -18,9 +19,8 @@ log = olympia.core.logger.getLogger('z.amo.blocklist')
 
 class MLBF():
     KEY_FORMAT = '{guid}:{version}'
-
-    blocked = []
-    not_blocked = []
+    # How many guids should there be in the stashs before we make a new base.
+    BASE_REPLACE_THRESHOLD = 500
 
     def __init__(self, id_):
         # simplify later code by assuming always a string
@@ -113,14 +113,24 @@ class MLBF():
             settings.MLBF_STORAGE_PATH, self.id, 'filter')
 
     @property
-    def blocked_path(self):
+    def _blocked_path(self):
         return os.path.join(
             settings.MLBF_STORAGE_PATH, self.id, 'blocked.json')
 
+    @cached_property
+    def blocked_json(self):
+        with storage.open(self._blocked_path, 'r') as json_file:
+            return json.load(json_file)
+
     @property
-    def not_blocked_path(self):
+    def _not_blocked_path(self):
         return os.path.join(
             settings.MLBF_STORAGE_PATH, self.id, 'notblocked.json')
+
+    @cached_property
+    def not_blocked_json(self):
+        with storage.open(self._not_blocked_path, 'r') as json_file:
+            return json.load(json_file)
 
     @property
     def stash_path(self):
@@ -137,12 +147,14 @@ class MLBF():
         else:
             version_excludes = ()
 
-        self.blocked = self.hash_filter_inputs(blocked)
-        self.not_blocked = self.hash_filter_inputs(
+        self.blocked_json = self.hash_filter_inputs(blocked)
+        self.not_blocked_json = self.hash_filter_inputs(
             not_blocked or self.get_all_guids(version_excludes))
 
         bloomfilter = self.generate_mlbf(
-            stats=stats, blocked=self.blocked, not_blocked=self.not_blocked)
+            stats=stats,
+            blocked=self.blocked_json,
+            not_blocked=self.not_blocked_json)
         # write bloomfilter
         mlbf_path = self.filter_path
         with storage.open(mlbf_path, 'wb') as filter_file:
@@ -150,15 +162,15 @@ class MLBF():
             bloomfilter.tofile(filter_file)
             stats['mlbf_filesize'] = os.stat(mlbf_path).st_size
         # write blocked json
-        blocked_path = self.blocked_path
+        blocked_path = self._blocked_path
         with storage.open(blocked_path, 'w') as json_file:
             log.info("Writing to file {}".format(blocked_path))
-            json.dump(self.blocked, json_file)
+            json.dump(self.blocked_json, json_file)
         # and the not blocked json
-        not_blocked_path = self.not_blocked_path
+        not_blocked_path = self._not_blocked_path
         with storage.open(not_blocked_path, 'w') as json_file:
             log.info("Writing to file {}".format(not_blocked_path))
-            json.dump(self.not_blocked, json_file)
+            json.dump(self.not_blocked_json, json_file)
 
         log.info(json.dumps(stats))
 
@@ -170,14 +182,10 @@ class MLBF():
         deletes = previous - current
         return extras, deletes
 
-    def write_stash(self, previous_id):
-        # load last mlbf blocked guids
-        previous_mlbf = MLBF(previous_id)
-        with storage.open(previous_mlbf.blocked_path, 'r') as json_file:
-            previous_blocked = json.load(json_file)
-        # compare with current blocks
+    def write_stash(self, previous_mlbf):
+        # compare previous with current blocks
         extras, deletes = self.generate_diffs(
-            previous_blocked, self.blocked)
+            previous_mlbf.blocked_json, self.blocked_json)
         stash = {
             'blocked': list(extras),
             'unblocked': list(deletes),
@@ -187,3 +195,13 @@ class MLBF():
         with storage.open(stash_path, 'w') as json_file:
             log.info("Writing to file {}".format(stash_path))
             json.dump(stash, json_file)
+
+    def should_reset_base_filter(self, previous_base_mlbf):
+        try:
+            # compare base with current blocks
+            extras, deletes = self.generate_diffs(
+                previous_base_mlbf.blocked_json, self.blocked_json)
+            return (len(extras) + len(deletes)) > self.BASE_REPLACE_THRESHOLD
+        except FileNotFoundError:
+            # when previous_base_mlfb._blocked_path doesn't exist
+            return True

--- a/src/olympia/blocklist/tasks.py
+++ b/src/olympia/blocklist/tasks.py
@@ -32,7 +32,7 @@ MLBF_TIME_CONFIG_KEY = 'blocklist_mlbf_generation_time'
 MLBF_BASE_ID_CONFIG_KEY = 'blocklist_mlbf_base_id'
 
 BLOCKLIST_RECORD_MLBF_BASE = 'bloomfilter-base'
-BLOCKLIST_RECORD_MLBF_UPDATE = 'bloomfilter-update'
+BLOCKLIST_RECORD_MLBF_UPDATE = 'bloomfilter-full'
 
 
 @task

--- a/src/olympia/blocklist/tasks.py
+++ b/src/olympia/blocklist/tasks.py
@@ -29,6 +29,11 @@ bracket_open_regex = re.compile(r'(?<!\\){')
 bracket_close_regex = re.compile(r'(?<!\\)}')
 
 MLBF_TIME_CONFIG_KEY = 'blocklist_mlbf_generation_time'
+MLBF_BASE_ID_CONFIG_KEY = 'blocklist_mlbf_base_id'
+
+BLOCKLIST_RECORD_MLBF_BASE = 'bloomfilter-base'
+BLOCKLIST_RECORD_MLBF_UPDATE = 'bloomfilter-update'
+BLOCKLIST_RECORD_STASH = 'stash'
 
 
 @task
@@ -143,16 +148,37 @@ def import_block_from_blocklist(record):
 
 
 @task
-def upload_filter_to_kinto(generation_time):
+def upload_filter_to_kinto(generation_time, is_base=True, upload_stash=False):
     server = KintoServer(
         KINTO_BUCKET, KINTO_COLLECTION_MLBF, kinto_sign_off_needed=False)
+    mlbf = MLBF(generation_time)
+    # Deal with possible stashes first
+    if upload_stash:
+        # If we have a stash, write the attachment for that
+        stash_data = {
+            'key_format': MLBF.KEY_FORMAT,
+            'stash_time': generation_time,
+            'attachment_type': BLOCKLIST_RECORD_STASH,
+        }
+        with storage.open(mlbf.stash_path, 'r') as stash_file:
+            attachment = ('stash.json', stash_file)
+            server.publish_attachment(stash_data, attachment)
+    if is_base:
+        # clear the collection for the base - we want to be the only filter
+        server.delete_all_records()
+
+    # Then the bloomfilter
     data = {
         'key_format': MLBF.KEY_FORMAT,
         'generation_time': generation_time,
+        'attachment_type':
+            BLOCKLIST_RECORD_MLBF_BASE if is_base else
+            BLOCKLIST_RECORD_MLBF_UPDATE,
     }
-    mlbf_path = MLBF(generation_time).filter_path
-    with storage.open(mlbf_path) as filter_file:
+    with storage.open(mlbf.filter_path, 'rb') as filter_file:
         attachment = ('filter.bin', filter_file, 'application/octet-stream')
         server.publish_attachment(data, attachment)
     server.complete_session()
     set_config(MLBF_TIME_CONFIG_KEY, generation_time, json_value=True)
+    if is_base:
+        set_config(MLBF_BASE_ID_CONFIG_KEY, generation_time, json_value=True)

--- a/src/olympia/blocklist/tasks.py
+++ b/src/olympia/blocklist/tasks.py
@@ -33,7 +33,6 @@ MLBF_BASE_ID_CONFIG_KEY = 'blocklist_mlbf_base_id'
 
 BLOCKLIST_RECORD_MLBF_BASE = 'bloomfilter-base'
 BLOCKLIST_RECORD_MLBF_UPDATE = 'bloomfilter-update'
-BLOCKLIST_RECORD_STASH = 'stash'
 
 
 @task
@@ -152,20 +151,18 @@ def upload_filter_to_kinto(generation_time, is_base=True, upload_stash=False):
     server = KintoServer(
         KINTO_BUCKET, KINTO_COLLECTION_MLBF, kinto_sign_off_needed=False)
     mlbf = MLBF(generation_time)
-    # Deal with possible stashes first
-    if upload_stash:
-        # If we have a stash, write the attachment for that
-        stash_data = {
-            'key_format': MLBF.KEY_FORMAT,
-            'stash_time': generation_time,
-            'attachment_type': BLOCKLIST_RECORD_STASH,
-        }
-        with storage.open(mlbf.stash_path, 'r') as stash_file:
-            attachment = ('stash.json', stash_file)
-            server.publish_attachment(stash_data, attachment)
     if is_base:
         # clear the collection for the base - we want to be the only filter
         server.delete_all_records()
+    # Deal with possible stashes first
+    if upload_stash:
+        # If we have a stash, write that
+        stash_data = {
+            'key_format': MLBF.KEY_FORMAT,
+            'stash_time': generation_time,
+            'stash': mlbf.stash_json,
+        }
+        server.publish_record(stash_data)
 
     # Then the bloomfilter
     data = {

--- a/src/olympia/blocklist/tests/test_cron.py
+++ b/src/olympia/blocklist/tests/test_cron.py
@@ -13,11 +13,14 @@ from olympia.amo.tests import addon_factory, TestCase, user_factory
 from olympia.blocklist.cron import upload_mlbf_to_kinto
 from olympia.blocklist.mlbf import MLBF
 from olympia.blocklist.models import Block
-from olympia.blocklist.tasks import MLBF_TIME_CONFIG_KEY
+from olympia.blocklist.tasks import (
+    MLBF_TIME_CONFIG_KEY, MLBF_BASE_ID_CONFIG_KEY)
 from olympia.lib.kinto import KintoServer
 from olympia.zadmin.models import get_config, set_config
 
 
+@freeze_time('2020-01-01 12:34:56')
+@override_switch('blocklist_mlbf_submit', active=True)
 class TestUploadToKinto(TestCase):
     def setUp(self):
         addon_factory()
@@ -25,44 +28,42 @@ class TestUploadToKinto(TestCase):
             addon=addon_factory(
                 file_kw={'is_signed': True, 'is_webextension': True}),
             updated_by=user_factory())
+        delete_patcher = mock.patch.object(KintoServer, 'delete_all_records')
+        publish_patcher = mock.patch.object(KintoServer, 'publish_attachment')
+        self.addCleanup(delete_patcher.stop)
+        self.addCleanup(publish_patcher.stop)
+        self.delete_mock = delete_patcher.start()
+        self.publish_mock = publish_patcher.start()
 
-    @freeze_time('2020-01-01 12:34:56')
-    @override_switch('blocklist_mlbf_submit', active=True)
-    @mock.patch.object(KintoServer, 'publish_attachment')
-    def test_upload_mlbf_to_kinto(self, publish_mock):
+    def test_no_previous_mlbf(self):
         upload_mlbf_to_kinto()
 
         generation_time = int(
             datetime.datetime(2020, 1, 1, 12, 34, 56).timestamp() * 1000)
-        publish_mock.assert_called_with(
+        self.publish_mock.assert_called_with(
             {'key_format': MLBF.KEY_FORMAT,
-             'generation_time': generation_time},
+             'generation_time': generation_time,
+             'attachment_type': 'bloomfilter-base'},
             ('filter.bin', mock.ANY, 'application/octet-stream'))
         assert (
             get_config(MLBF_TIME_CONFIG_KEY, json_value=True) ==
             generation_time)
+        assert (
+            get_config(MLBF_BASE_ID_CONFIG_KEY, json_value=True) ==
+            generation_time)
+        self.delete_mock.assert_called_once()
 
-        mlfb_path = os.path.join(
-            settings.MLBF_STORAGE_PATH, str(generation_time), 'filter')
-        assert os.path.exists(mlfb_path)
-        assert os.path.getsize(mlfb_path)
+        gen_path = os.path.join(
+            settings.MLBF_STORAGE_PATH, str(generation_time))
+        assert os.path.getsize(os.path.join(gen_path, 'filter'))
+        assert os.path.getsize(os.path.join(gen_path, 'blocked.json'))
+        assert os.path.getsize(os.path.join(gen_path, 'notblocked.json'))
+        # no stash because no previous mlbf
+        assert not os.path.exists(os.path.join(gen_path, 'stash.json'))
 
-        blocked_path = os.path.join(
-            settings.MLBF_STORAGE_PATH, str(generation_time), 'blocked.json')
-        assert os.path.exists(blocked_path)
-        assert os.path.getsize(blocked_path)
-
-        not_blocked_path = os.path.join(
-            settings.MLBF_STORAGE_PATH, str(generation_time),
-            'notblocked.json')
-        assert os.path.exists(not_blocked_path)
-        assert os.path.getsize(not_blocked_path)
-
-    @freeze_time('2020-01-01 12:34:56')
-    @override_switch('blocklist_mlbf_submit', active=True)
-    @mock.patch.object(KintoServer, 'publish_attachment')
-    def test_stash_file(self, publish_mock):
+    def test_stash_because_previous_mlbf(self):
         set_config(MLBF_TIME_CONFIG_KEY, 123456, json_value=True)
+        set_config(MLBF_BASE_ID_CONFIG_KEY, 123456, json_value=True)
         prev_blocked_path = os.path.join(
             settings.MLBF_STORAGE_PATH, '123456', 'blocked.json')
         with storage.open(prev_blocked_path, 'w') as blocked_file:
@@ -73,9 +74,21 @@ class TestUploadToKinto(TestCase):
         generation_time = int(
             datetime.datetime(2020, 1, 1, 12, 34, 56).timestamp() * 1000)
 
+        self.publish_mock.assert_called_with(
+            {'key_format': MLBF.KEY_FORMAT,
+             'generation_time': generation_time,
+             'attachment_type': 'bloomfilter-update'},
+            ('filter.bin', mock.ANY, 'application/octet-stream'))
+        self.delete_mock.assert_not_called()
+        assert (
+            get_config(MLBF_TIME_CONFIG_KEY, json_value=True) ==
+            generation_time)
+        assert (
+            get_config(MLBF_BASE_ID_CONFIG_KEY, json_value=True) ==
+            123456)
+
         stash_path = os.path.join(
             settings.MLBF_STORAGE_PATH, str(generation_time), 'stash.json')
-        assert os.path.exists(stash_path)
         assert os.path.getsize(stash_path)
         with open(stash_path) as stash_file:
             blocked_guid = (
@@ -85,18 +98,92 @@ class TestUploadToKinto(TestCase):
                 'blocked': [blocked_guid],
                 'unblocked': ['madeup@guid:123']}
 
-    @override_switch('blocklist_mlbf_submit', active=False)
-    @mock.patch.object(KintoServer, 'publish_attachment')
-    def test_waffle_off_disables_publishing(self, publish_mock):
+    def test_stash_because_many_mlbf(self):
+        set_config(MLBF_TIME_CONFIG_KEY, 123456, json_value=True)
+        set_config(MLBF_BASE_ID_CONFIG_KEY, 987654, json_value=True)
+        prev_blocked_path = os.path.join(
+            settings.MLBF_STORAGE_PATH, '123456', 'blocked.json')
+        with storage.open(prev_blocked_path, 'w') as blocked_file:
+            json.dump(['madeup@guid:12345'], blocked_file)
+        base_blocked_path = os.path.join(
+            settings.MLBF_STORAGE_PATH, '987654', 'blocked.json')
+        with storage.open(base_blocked_path, 'w') as blocked_file:
+            json.dump([], blocked_file)
+
         upload_mlbf_to_kinto()
 
-        publish_mock.assert_not_called()
+        generation_time = int(
+            datetime.datetime(2020, 1, 1, 12, 34, 56).timestamp() * 1000)
+
+        self.publish_mock.assert_called_with(
+            {'key_format': MLBF.KEY_FORMAT,
+             'generation_time': generation_time,
+             'attachment_type': 'bloomfilter-update'},
+            ('filter.bin', mock.ANY, 'application/octet-stream'))
+        self.delete_mock.assert_not_called()
+        assert (
+            get_config(MLBF_TIME_CONFIG_KEY, json_value=True) ==
+            generation_time)
+        assert (
+            get_config(MLBF_BASE_ID_CONFIG_KEY, json_value=True) ==
+            987654)
+
+        stash_path = os.path.join(
+            settings.MLBF_STORAGE_PATH, str(generation_time), 'stash.json')
+        assert os.path.getsize(stash_path)
+        with open(stash_path) as stash_file:
+            blocked_guid = (
+                f'{self.block.guid}:'
+                f'{self.block.addon.current_version.version}')
+            assert json.load(stash_file) == {
+                'blocked': [blocked_guid],
+                'unblocked': ['madeup@guid:12345']}
+
+    @mock.patch.object(MLBF, 'should_reset_base_filter')
+    def test_reset_base_because_over_reset_threshold(self, should_reset_mock):
+        should_reset_mock.return_value = True
+        set_config(MLBF_TIME_CONFIG_KEY, 123456, json_value=True)
+        set_config(MLBF_BASE_ID_CONFIG_KEY, 987654, json_value=True)
+        prev_blocked_path = os.path.join(
+            settings.MLBF_STORAGE_PATH, '123456', 'blocked.json')
+        with storage.open(prev_blocked_path, 'w') as blocked_file:
+            json.dump(['madeup@guid:12345'], blocked_file)
+        base_blocked_path = os.path.join(
+            settings.MLBF_STORAGE_PATH, '987654', 'blocked.json')
+        with storage.open(base_blocked_path, 'w') as blocked_file:
+            json.dump([], blocked_file)
+
+        upload_mlbf_to_kinto()
+
+        generation_time = int(
+            datetime.datetime(2020, 1, 1, 12, 34, 56).timestamp() * 1000)
+
+        self.publish_mock.assert_called_with(
+            {'key_format': MLBF.KEY_FORMAT,
+             'generation_time': generation_time,
+             'attachment_type': 'bloomfilter-base'},
+            ('filter.bin', mock.ANY, 'application/octet-stream'))
+        self.delete_mock.assert_called_once()
+        assert (
+            get_config(MLBF_TIME_CONFIG_KEY, json_value=True) ==
+            generation_time)
+        assert (
+            get_config(MLBF_BASE_ID_CONFIG_KEY, json_value=True) ==
+            generation_time)
+
+        gen_path = os.path.join(
+            settings.MLBF_STORAGE_PATH, str(generation_time))
+        # no stash because we're starting with a new base mlbf
+        assert not os.path.exists(os.path.join(gen_path, 'stash.json'))
+
+    @override_switch('blocklist_mlbf_submit', active=False)
+    def test_waffle_off_disables_publishing(self):
+        upload_mlbf_to_kinto()
+
+        self.publish_mock.assert_not_called()
         assert not get_config(MLBF_TIME_CONFIG_KEY)
 
-    @freeze_time('2020-01-01 12:34:56')
-    @override_switch('blocklist_mlbf_submit', active=True)
-    @mock.patch.object(KintoServer, 'publish_attachment')
-    def test_no_need_for_new_mlbf(self, publish_mock):
+    def test_no_block_changes(self):
         # This was the last time the mlbf was generated
         last_time = int(
             datetime.datetime(2020, 1, 1, 12, 34, 1).timestamp() * 1000)
@@ -105,7 +192,7 @@ class TestUploadToKinto(TestCase):
         set_config(MLBF_TIME_CONFIG_KEY, last_time, json_value=True)
         upload_mlbf_to_kinto()
         # So no need for a new bloomfilter
-        publish_mock.assert_not_called()
+        self.publish_mock.assert_not_called()
 
         # But if we add a new Block a new filter is needed
         addon_factory()
@@ -114,7 +201,7 @@ class TestUploadToKinto(TestCase):
                 file_kw={'is_signed': True, 'is_webextension': True}),
             updated_by=user_factory())
         upload_mlbf_to_kinto()
-        publish_mock.assert_called_once()
+        self.publish_mock.assert_called_once()
         assert (
             get_config(MLBF_TIME_CONFIG_KEY, json_value=True) ==
             int(datetime.datetime(2020, 1, 1, 12, 34, 56).timestamp() * 1000))

--- a/src/olympia/blocklist/tests/test_cron.py
+++ b/src/olympia/blocklist/tests/test_cron.py
@@ -81,7 +81,7 @@ class TestUploadToKinto(TestCase):
         self.publish_attachment_mock.assert_called_with(
             {'key_format': MLBF.KEY_FORMAT,
              'generation_time': generation_time,
-             'attachment_type': 'bloomfilter-update'},
+             'attachment_type': 'bloomfilter-full'},
             ('filter.bin', mock.ANY, 'application/octet-stream'))
         self.publish_record_mock.assert_called_with({
             'key_format': MLBF.KEY_FORMAT,
@@ -120,7 +120,7 @@ class TestUploadToKinto(TestCase):
         self.publish_attachment_mock.assert_called_with(
             {'key_format': MLBF.KEY_FORMAT,
              'generation_time': generation_time,
-             'attachment_type': 'bloomfilter-update'},
+             'attachment_type': 'bloomfilter-full'},
             ('filter.bin', mock.ANY, 'application/octet-stream'))
         self.publish_record_mock.assert_called_with({
             'key_format': MLBF.KEY_FORMAT,

--- a/src/olympia/blocklist/tests/test_cron.py
+++ b/src/olympia/blocklist/tests/test_cron.py
@@ -29,18 +29,21 @@ class TestUploadToKinto(TestCase):
                 file_kw={'is_signed': True, 'is_webextension': True}),
             updated_by=user_factory())
         delete_patcher = mock.patch.object(KintoServer, 'delete_all_records')
-        publish_patcher = mock.patch.object(KintoServer, 'publish_attachment')
+        attach_patcher = mock.patch.object(KintoServer, 'publish_attachment')
+        record_patcher = mock.patch.object(KintoServer, 'publish_record')
         self.addCleanup(delete_patcher.stop)
-        self.addCleanup(publish_patcher.stop)
+        self.addCleanup(attach_patcher.stop)
+        self.addCleanup(record_patcher.stop)
         self.delete_mock = delete_patcher.start()
-        self.publish_mock = publish_patcher.start()
+        self.publish_attachment_mock = attach_patcher.start()
+        self.publish_record_mock = record_patcher.start()
 
     def test_no_previous_mlbf(self):
         upload_mlbf_to_kinto()
 
         generation_time = int(
             datetime.datetime(2020, 1, 1, 12, 34, 56).timestamp() * 1000)
-        self.publish_mock.assert_called_with(
+        self.publish_attachment_mock.assert_called_with(
             {'key_format': MLBF.KEY_FORMAT,
              'generation_time': generation_time,
              'attachment_type': 'bloomfilter-base'},
@@ -51,6 +54,7 @@ class TestUploadToKinto(TestCase):
         assert (
             get_config(MLBF_BASE_ID_CONFIG_KEY, json_value=True) ==
             generation_time)
+        self.publish_record_mock.assert_not_called()
         self.delete_mock.assert_called_once()
 
         gen_path = os.path.join(
@@ -74,11 +78,20 @@ class TestUploadToKinto(TestCase):
         generation_time = int(
             datetime.datetime(2020, 1, 1, 12, 34, 56).timestamp() * 1000)
 
-        self.publish_mock.assert_called_with(
+        self.publish_attachment_mock.assert_called_with(
             {'key_format': MLBF.KEY_FORMAT,
              'generation_time': generation_time,
              'attachment_type': 'bloomfilter-update'},
             ('filter.bin', mock.ANY, 'application/octet-stream'))
+        self.publish_record_mock.assert_called_with({
+            'key_format': MLBF.KEY_FORMAT,
+            'stash_time': generation_time,
+            'stash': {
+                'blocked': [
+                    f'{self.block.guid}:'
+                    f'{self.block.addon.current_version.version}'],
+                'unblocked': ['madeup@guid:123']}
+        })
         self.delete_mock.assert_not_called()
         assert (
             get_config(MLBF_TIME_CONFIG_KEY, json_value=True) ==
@@ -86,17 +99,6 @@ class TestUploadToKinto(TestCase):
         assert (
             get_config(MLBF_BASE_ID_CONFIG_KEY, json_value=True) ==
             123456)
-
-        stash_path = os.path.join(
-            settings.MLBF_STORAGE_PATH, str(generation_time), 'stash.json')
-        assert os.path.getsize(stash_path)
-        with open(stash_path) as stash_file:
-            blocked_guid = (
-                f'{self.block.guid}:'
-                f'{self.block.addon.current_version.version}')
-            assert json.load(stash_file) == {
-                'blocked': [blocked_guid],
-                'unblocked': ['madeup@guid:123']}
 
     def test_stash_because_many_mlbf(self):
         set_config(MLBF_TIME_CONFIG_KEY, 123456, json_value=True)
@@ -115,11 +117,20 @@ class TestUploadToKinto(TestCase):
         generation_time = int(
             datetime.datetime(2020, 1, 1, 12, 34, 56).timestamp() * 1000)
 
-        self.publish_mock.assert_called_with(
+        self.publish_attachment_mock.assert_called_with(
             {'key_format': MLBF.KEY_FORMAT,
              'generation_time': generation_time,
              'attachment_type': 'bloomfilter-update'},
             ('filter.bin', mock.ANY, 'application/octet-stream'))
+        self.publish_record_mock.assert_called_with({
+            'key_format': MLBF.KEY_FORMAT,
+            'stash_time': generation_time,
+            'stash': {
+                'blocked': [
+                    f'{self.block.guid}:'
+                    f'{self.block.addon.current_version.version}'],
+                'unblocked': ['madeup@guid:12345']}
+        })
         self.delete_mock.assert_not_called()
         assert (
             get_config(MLBF_TIME_CONFIG_KEY, json_value=True) ==
@@ -127,17 +138,6 @@ class TestUploadToKinto(TestCase):
         assert (
             get_config(MLBF_BASE_ID_CONFIG_KEY, json_value=True) ==
             987654)
-
-        stash_path = os.path.join(
-            settings.MLBF_STORAGE_PATH, str(generation_time), 'stash.json')
-        assert os.path.getsize(stash_path)
-        with open(stash_path) as stash_file:
-            blocked_guid = (
-                f'{self.block.guid}:'
-                f'{self.block.addon.current_version.version}')
-            assert json.load(stash_file) == {
-                'blocked': [blocked_guid],
-                'unblocked': ['madeup@guid:12345']}
 
     @mock.patch.object(MLBF, 'should_reset_base_filter')
     def test_reset_base_because_over_reset_threshold(self, should_reset_mock):
@@ -158,11 +158,12 @@ class TestUploadToKinto(TestCase):
         generation_time = int(
             datetime.datetime(2020, 1, 1, 12, 34, 56).timestamp() * 1000)
 
-        self.publish_mock.assert_called_with(
+        self.publish_attachment_mock.assert_called_with(
             {'key_format': MLBF.KEY_FORMAT,
              'generation_time': generation_time,
              'attachment_type': 'bloomfilter-base'},
             ('filter.bin', mock.ANY, 'application/octet-stream'))
+        self.publish_record_mock.assert_not_called()
         self.delete_mock.assert_called_once()
         assert (
             get_config(MLBF_TIME_CONFIG_KEY, json_value=True) ==
@@ -180,7 +181,8 @@ class TestUploadToKinto(TestCase):
     def test_waffle_off_disables_publishing(self):
         upload_mlbf_to_kinto()
 
-        self.publish_mock.assert_not_called()
+        self.publish_attachment_mock.assert_not_called()
+        self.publish_record_mock.assert_not_called()
         assert not get_config(MLBF_TIME_CONFIG_KEY)
 
     def test_no_block_changes(self):
@@ -192,7 +194,8 @@ class TestUploadToKinto(TestCase):
         set_config(MLBF_TIME_CONFIG_KEY, last_time, json_value=True)
         upload_mlbf_to_kinto()
         # So no need for a new bloomfilter
-        self.publish_mock.assert_not_called()
+        self.publish_attachment_mock.assert_not_called()
+        self.publish_record_mock.assert_not_called()
 
         # But if we add a new Block a new filter is needed
         addon_factory()
@@ -201,7 +204,7 @@ class TestUploadToKinto(TestCase):
                 file_kw={'is_signed': True, 'is_webextension': True}),
             updated_by=user_factory())
         upload_mlbf_to_kinto()
-        self.publish_mock.assert_called_once()
+        self.publish_attachment_mock.assert_called_once()
         assert (
             get_config(MLBF_TIME_CONFIG_KEY, json_value=True) ==
             int(datetime.datetime(2020, 1, 1, 12, 34, 56).timestamp() * 1000))

--- a/src/olympia/blocklist/tests/test_mlbf.py
+++ b/src/olympia/blocklist/tests/test_mlbf.py
@@ -1,6 +1,7 @@
 import json
 import os
 import tempfile
+from unittest import mock
 
 from filtercascade import FilterCascade
 
@@ -57,6 +58,7 @@ class TestMLBF(TestCase):
         # no matching versions (edge cases)
         self.over = Block.objects.create(
             addon=addon_factory(
+                version_kw={'version': '0.1'},
                 file_kw={'is_signed': True, 'is_webextension': True}),
             updated_by=user_factory(),
             max_version='0')
@@ -194,7 +196,7 @@ class TestMLBF(TestCase):
         old_mlbf.generate_and_write_mlbf()
         new_mlbf = MLBF('new_no_change')
         new_mlbf.generate_and_write_mlbf()
-        new_mlbf.write_stash('old')
+        new_mlbf.write_stash(old_mlbf)
         with open(new_mlbf.stash_path) as stash_file:
             assert json.load(stash_file) == {'blocked': [], 'unblocked': []}
         # add a new Block and delete one
@@ -207,8 +209,42 @@ class TestMLBF(TestCase):
         self.five_ver_123_5.delete(hard=True)
         newer_mlbf = MLBF('new_one_change')
         newer_mlbf.generate_and_write_mlbf()
-        newer_mlbf.write_stash('new_no_change')
+        newer_mlbf.write_stash(new_mlbf)
         with open(newer_mlbf.stash_path) as stash_file:
             assert json.load(stash_file) == {
                 'blocked': ['fooo@baaaa:999'],
                 'unblocked': [f'{self.five_ver.guid}:123.5']}
+
+    def test_should_reset_base_filter(self):
+        base_mlbf = MLBF('base')
+        base_mlbf.generate_and_write_mlbf()
+        # should handle the files not existing
+        assert MLBF('no_files').should_reset_base_filter(base_mlbf)
+
+        no_change_mlbf = MLBF('no_change')
+        no_change_mlbf.generate_and_write_mlbf()
+        assert not no_change_mlbf.should_reset_base_filter(base_mlbf)
+
+        # make some changes
+        Block.objects.create(
+            addon=addon_factory(
+                guid='fooo@baaaa',
+                version_kw={'version': '999'},
+                file_kw={'is_signed': True, 'is_webextension': True}),
+            updated_by=user_factory())
+        self.five_ver_123_5.delete(hard=True)
+        small_change_mlbf = MLBF('small_change')
+        small_change_mlbf.generate_and_write_mlbf()
+        # but the changes were small (less than threshold) so no need for new
+        # base filter
+        assert not small_change_mlbf.should_reset_base_filter(base_mlbf)
+        # double check what the differences were
+        diffs = MLBF.generate_diffs(
+            previous=base_mlbf.blocked_json,
+            current=small_change_mlbf.blocked_json)
+        assert diffs == ({'fooo@baaaa:999'}, {f'{self.five_ver.guid}:123.5'})
+
+        # so lower the threshold
+        to_patch = 'olympia.blocklist.mlbf.MLBF.BASE_REPLACE_THRESHOLD'
+        with mock.patch(to_patch, 1):
+            assert small_change_mlbf.should_reset_base_filter(base_mlbf)

--- a/src/olympia/blocklist/tests/test_mlbf.py
+++ b/src/olympia/blocklist/tests/test_mlbf.py
@@ -197,8 +197,10 @@ class TestMLBF(TestCase):
         new_mlbf = MLBF('new_no_change')
         new_mlbf.generate_and_write_mlbf()
         new_mlbf.write_stash(old_mlbf)
+        empty_stash = {'blocked': [], 'unblocked': []}
         with open(new_mlbf.stash_path) as stash_file:
-            assert json.load(stash_file) == {'blocked': [], 'unblocked': []}
+            assert json.load(stash_file) == empty_stash
+        assert new_mlbf.stash_json == empty_stash
         # add a new Block and delete one
         Block.objects.create(
             addon=addon_factory(
@@ -210,10 +212,12 @@ class TestMLBF(TestCase):
         newer_mlbf = MLBF('new_one_change')
         newer_mlbf.generate_and_write_mlbf()
         newer_mlbf.write_stash(new_mlbf)
+        full_stash = {
+            'blocked': ['fooo@baaaa:999'],
+            'unblocked': [f'{self.five_ver.guid}:123.5']}
         with open(newer_mlbf.stash_path) as stash_file:
-            assert json.load(stash_file) == {
-                'blocked': ['fooo@baaaa:999'],
-                'unblocked': [f'{self.five_ver.guid}:123.5']}
+            assert json.load(stash_file) == full_stash
+        assert newer_mlbf.stash_json == full_stash
 
     def test_should_reset_base_filter(self):
         base_mlbf = MLBF('base')

--- a/src/olympia/lib/kinto.py
+++ b/src/olympia/lib/kinto.py
@@ -160,6 +160,14 @@ class KintoServer(object):
             url, headers=self.headers)
         self._changes = True
 
+    def delete_all_records(self):
+        self.setup()
+        url = (
+            f'{settings.KINTO_API_URL}buckets/{self.bucket}/'
+            f'collections/{self.collection}/records')
+        requests.delete(url, headers=self.headers)
+        self._changes = True
+
     def complete_session(self):
         if not self._changes:
             return

--- a/src/olympia/lib/tests/test_kinto.py
+++ b/src/olympia/lib/tests/test_kinto.py
@@ -191,6 +191,21 @@ class TestKintoServer(TestCase):
         server.delete_record('an-id')
         assert server._changes
 
+    def test_delete_all_records(self):
+        server = KintoServer('foo', 'baa')
+        server._setup_done = True
+        assert not server._changes
+        url = (
+            settings.KINTO_API_URL +
+            'buckets/foo/collections/baa/records')
+        responses.add(
+            responses.DELETE,
+            url,
+            content_type='application/json')
+
+        server.delete_all_records()
+        assert server._changes
+
     def test_complete_session(self):
         server = KintoServer('foo', 'baa')
         server._setup_done = True

